src/common/meta/src/kv_backend/postgres.rs

```diff
@@ -0,0 +1,834 @@
+                true => sql.push_str(&format!(" WHERE k like '{start_key}%'")),
+                false => {
+                    sql.push_str(&format!(" WHERE k >= '{start_key}'"));
+                    if req.range_end != b"\0" {
+                        let end_key = process_bytes(req.range_end, "DeleteRangeRequestRangeEnd")?;
+                        sql.push_str(&format!(" AND k <= '{end_key}'"));
+                    }
+                }
+            },
+        }
+    }
+    sql.push_str(" RETURNING k, v");
+    sql.push(';');
+    Ok(sql)
+}
+
+/*
+DELETE FROM metakv WHERE k IN (kvs) RETURNING K, V;
+*/
+fn generate_batch_delete_query(req: BatchDeleteRequest) -> Result<String> {
+    let keys = req
+        .keys
+        .iter()
+        .map(|k| process_bytes(k.clone(), "BatchDeleteRequestKey"))
+        .collect::<Result<Vec<String>>>()?;
+    let mut sql = String::new();
+    sql.push_str("DELETE FROM metakv ");
+    let predicate = keys
+        .iter()
+        .map(|k| format!("'{k}'"))
+        .collect::<Vec<String>>()
+        .join(",");
+    sql.push_str(&format!("WHERE k IN ({predicate})"));
+    sql.push_str(" RETURNING k, v");
+    sql.push(';');
+    Ok(sql)
+}
+
+/*
+WITH prev AS (
+        select k,v from metakv where k = '{key}' AND v = '{expect}'
+    ), update AS (
+    UPDATE metakv
+    SET k='{key}',
+    v='{value}'
+    WHERE
+        k='{key}'
+        AND v='{expected}'
+    )
+
+    SELECT k, v FROM prev;
+*/
+fn generate_compare_and_put_query(req: CompareAndPutRequest) -> Result<String> {
+    let key = process_bytes(req.key, "CompareAndPutRequestKey")?;
+    let expect = process_bytes(req.expect, "CompareAndPutRequestExpect")?;
+    let value = process_bytes(req.value, "CompareAndPutRequestValue")?;
+
+    Ok(format!(
+        r#"
+    WITH prev AS (
+        select k,v from metakv where k = '{key}' AND v = '{expect}'
+    ), update AS (
+    UPDATE metakv
+    SET k='{key}',
+    v='{value}'
+    WHERE 
+        k='{key}' AND v = '{expect}'
+    )
+
+    SELECT k, v FROM prev;
+    "#
+    ))
+}
+
+fn generate_put_if_not_exist_query(key: Vec<u8>, value: Vec<u8>) -> Result<String> {
+    let key = process_bytes(key, "PutIfNotExistKey")?;
+    let value = process_bytes(value, "PutIfNotExistValue")?;
+    Ok(format!(
+        r#"    
+    WITH prev AS (
+        select k,v from metakv where k = '{key}'
+    ), insert AS (
+        INSERT INTO metakv
+        VALUES ('{key}', '{value}')
+        ON CONFLICT (k) DO NOTHING
+    )
+
+    SELECT k, v FROM prev;"#
+    ))
+}
+
+//  Trim null byte at the end and convert bytes to string.
+fn process_bytes(mut data: Vec<u8>, name: &str) -> Result<String> {
+    let mut len = data.len();
+    // remove trailing null bytes to avoid error in postgres encoding.
+    while len > 0 && data[len - 1] == 0 {
+        len -= 1;
+    }
+    data.truncate(len);
+    let res = String::from_utf8(data).context(FromUtf8Snafu { name })?;
+    Ok(res)
+}
```

Comment:
```suggestion
fn process_bytes<'a>(data: &'a [u8], name: &str) -> Result<&'a str> {
    let mut len = data.len();
    // remove trailing null bytes to avoid error in postgres encoding.
    while len > 0 && data[len - 1] == 0 {
        len -= 1;
    }
    let res = str::from_utf8(&data[0..len]).context(FromUtf8Snafu { name })?;
    Ok(res)
}
```

Reply:
the tricky part is `FromUtf8Snafu` takes `error: std::string::FromUtf8Error` as source is it okay if i add a new error that take `error: std::str::Utf8Error` as source?